### PR TITLE
Load schemas in separate resource from State

### PR DESCRIPTION
**Pull request by oliverguenther**

### CODE REVIEW

File: frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts [hunk 1]
```diff
@@ -50,7 +52,7 @@ interface WorkPackageResourceEmbedded {
   project: HalResource|any;
   relations: CollectionResourceInterface;
   responsible: HalResource|any;
-  schema: HalResource|any;
+  schema: SchemaResource|any;
```
Comment (ulferts) [inline frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts]: Is it possible to remove the `|any` declaration?

File: frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts [hunk 2]
```diff
@@ -146,6 +149,10 @@ export class WorkPackageResource extends HalResource {
     return this.modifiedFields.length > 0;
   }
 
+  /**
+   * Returns a reference to the
+   */
```
Comment (ulferts) [inline frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts:155]: Something is missing here

File: frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts [hunk 3]
```diff
@@ -319,14 +310,12 @@ export class WorkPackageResource extends HalResource {
...
-        // Override the current schema with
-        // the changes from API
-        this.schema = form.$embedded.schema;
-
         // Take over new values from the form
         // this resource doesn't know yet.
         this.assignNewValues(form.$embedded.payload);
 
+        this.schema = form.$embedded.schema;
+
```
Comment (ulferts) [inline frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts]: Why this change?

File: frontend/app/components/work-packages/work-package-cache.service.ts [hunk 4]
```diff
@@ -74,8 +83,10 @@ export class WorkPackageCacheService {
       state.clear();
     }
 
-    state.putFromPromiseIfPristine(
-      () => this.apiWorkPackages.loadWorkPackageById(workPackageId, forceUpdate));
+    if (workPackageId.toString() !== 'new') {
+    state.putFromPromiseIfPristine(() =>
+        this.apiWorkPackages.loadWorkPackageById(workPackageId, forceUpdate));
+    }
```
Comment (ulferts) [inline frontend/app/components/work-packages/work-package-cache.service.ts:93]: The responsibilities between `WorkPackageCacheService` and `ApiWorkPackageService` are not well separated. For `list`, the schemas are loaded by the `WorkPackageCache` while the `ApiWorkPackageService` loads the schema if `loadWorkPackage` is called: https://github.com/opf/openproject/pull/5063/files#diff-030dc6d5af01418720e8f296bf3b048aR65

File: frontend/app/components/api/api-v3/hal-resources/hal-resource.service.ts [hunk 5]
```diff
@@ -100,7 +101,27 @@ export class HalResource {
     this.$initialize($source);
   }
 
+  /**
+   * Return the associated state to this HAL resource, if any.
+   */
+  public get state():State<HalResource>|null {
+    return null;
```
Comment (ulferts) [inline frontend/app/components/api/api-v3/hal-resources/hal-resource.service.ts:108]: I like the line of thought of each object having a reference to it's state and it looks nice in the `SchemaResource`. But it is not used in the WorkPackageResource, I believe because of the `WorkPackageCacheService` which is inconsistent IMO.

**Review by ulferts — CHANGES_REQUESTED** [review]
The change works and I like that it was possible to remove that much code.

What irks me though, maybe because it's been a while since I took a look at the code, is the mix of responsibilities between a Resource, its State and the service responsible for loading.

File: frontend/app/components/api/api-v3/hal-resources/hal-resource.service.ts (hunk 5 above)
Comment (oliverguenther) [reply frontend/app/components/api/api-v3/hal-resources/hal-resource.service.ts:108]: I agree. For sake of consistency, I have added the state to the work package resource, even though it is not in use by the work package. We cannot use that particular method inside  `#loadWorkPackage` anyway.

My next goal would be to remove the WP cache service. We can instead create a compound state of WorkPackage and Schema that is updated whenever each state is updated. That would be a rather big change due to the numerous references to the cache service.

**Comment by oliverguenther** [comment]
I've addressed most your remarks. I won't be able to resolve the increased responsibility of the cache service now without refactoring all other usages to it.

File: frontend/app/components/api/api-v3/hal-resources/hal-resource.service.ts (hunk 5 above)
Comment (oliverguenther) [reply frontend/app/components/api/api-v3/hal-resources/hal-resource.service.ts:108]: I had to remove the state for work packages again, since there is no process that loads the work packages if e.g., a child is added that is not yet loaded in the state. We have to refactor the cache service before re-introducing it.

**Review by ulferts — APPROVED** [review]
